### PR TITLE
feat(new-buffer) : changes acc. to no new buffer eslint rule

### PR DESCRIPTION
**Pull request by divyabiyani**
Changes according to mozilla/eslint-plugin-fxa#2.
Connects with the issue mozilla/eslint-plugin-fxa#1.

Fixes https://github.com/mozilla/fxa-auth-server/issues/1612

### CODE REVIEW

**Comment by seanmonstar** [comment]
Customs lints feel great, huh? :D

**Comment by vladikoff** [comment]
Fixes https://github.com/mozilla/fxa-auth-server/issues/1612

**Comment by vladikoff** [comment]
@divyabiyani 

```
[09:23:51]  <vladikoff>	so for this PR https://github.com/mozilla/fxa-auth-server/pull/1624/files
[09:23:58]  <vladikoff>	because it is a change with `Buffer`
[09:24:12]  <vladikoff>	i feel like we should do a test run with a custom branch
```

Make a change to package.json and set the source in
https://github.com/mozilla/fxa-auth-server/blob/master/package.json#L69

to

`"eslint-plugin-fxa": "git+https://github.com/divyabiyani/eslint-plugin-fxa#buffer",`
and run `npm run shrinkwrap` commit the changes and push to github, this way we are going to run Travis against that branch. If tests all pass on Travis then I will merge this change https://github.com/mozilla/eslint-plugin-fxa/pull/2/files and we will update this commit to reference the new SHA.

File: lib/crypto/base32.js [hunk 1]
```diff
@@ -21,7 +21,7 @@ function base32(len) {
     .then(bytes => {
       const out = []
 
-      for (let i = 0; i < len; i++) {
+      for (let i = 0; i < len; i ++) {
```
Comment (vladikoff) [inline lib/crypto/base32.js]: @divyabiyani these changes need to be part of a new pull request, both in this repo and in the eslint-plugin-fxa repo.

I reopened and cherry picked it here and waiting for the tests to pass: https://github.com/mozilla/fxa-auth-server/pull/1636

File: lib/customs.js [hunk 2]
```diff
@@ -61,7 +61,7 @@ module.exports = function (log, error) {
           // log a flow event that user got blocked.
           request.emitMetricsEvent('customs.blocked')
 
-          var unblock = !!result.unblock
+          var unblock = !! result.unblock
```
Comment (vladikoff) [inline lib/customs.js]: @divyabiyani 👍

File: lib/crypto/butil.js [hunk 3]
```diff
@@ -35,7 +35,7 @@ module.exports.xorBuffers = function xorBuffers(buffer1, buffer2) {
 module.exports.unbuffer = function unbuffer(object, inplace) {
   var keys = Object.keys(object)
   var copy = inplace ? object : {}
-  for (var i = 0; i < keys.length; i++) {
+  for (var i = 0; i < keys.length; i ++) {
```
Comment (vladikoff) [inline lib/crypto/butil.js]: @divyabiyani is there a way we can keep `i++`, but have `!! result.unblock`? (no spaces between `i` and `++`) ?

File: lib/crypto/base32.js (hunk 1 above)
Comment (vladikoff) [inline lib/crypto/base32.js]: @divyabiyani so let's reopen a new PR, we can also include this rule in the new PR: https://github.com/pdehaan/eslint-config-fxa/pull/11/files

**Comment by vladikoff** [comment]
@divyabiyani see https://github.com/mozilla/fxa-auth-server/pull/1624#discussion_r98958100 

let's open a new PR for this...

File: lib/crypto/butil.js (hunk 3 above)
Comment (divyabiyani) [inline lib/crypto/butil.js]: @vladikoff : What about ++i ?

Comment (pdehaan) [inline lib/crypto/butil.js]: @vladikoff: I _think_ the ESLint rule you're looking for is [`space-unary-ops`](http://eslint.org/docs/rules/space-unary-ops); plus you'll need some overrides for `!` and `!!`:

```yaml
env:
  es6: true
  node: true

extends:
  - eslint:recommended

root: true

rules:
  no-console: off
  space-unary-ops: [error, {words: true, nonwords: false, overrides: {'!': true, '!!': true}}]
```

That should allow the following rough styles:

```js
function foo(a, b) {
  a = a++;
  b = ++b;
  return !! a || !! b;
}

const res = foo(3, 4);
console.log(res);
```

If you're lucky,  you can run `eslint . --fix` and have it magically autofix+format all the code for you, but that may cause a slightly insane diff, depending on current config.

Comment (seanmonstar) [inline lib/crypto/butil.js]: Oh look, a bikeshed! I personally hate the space in unary operators, pluses and bangs.

Comment (vladikoff) [inline lib/crypto/butil.js]: @shane-tomlinson @philbooth I'm away today, could you please add your input on this and discuss it with @divyabiayni, so we can land this PR. Running the rule against the content server code base could be useful

Comment (vladikoff) [inline lib/crypto/butil.js]: Following Firefox desktop style, my preferences are:

1) `i++` and `++i`
2) `if (! thing)`

Comment (shane-tomlinson) [inline lib/crypto/butil.js]: My style preferences match @vladikoff's.

Comment (shane-tomlinson) [inline lib/crypto/butil.js]: Using @pdehaan's rule (modified to add quotes around error), the content server passes!

```
'space-unary-ops': ['error', {words: true, nonwords: false, overrides: {'!': true, '!!': true}}],
```

```bash
➜  fxa-content-server git:(master) grunt lint                                       git:(master|)
Running "eslint:app" (eslint) task

Running "eslint:grunt" (eslint) task

Running "eslint:scripts" (eslint) task

Running "eslint:server" (eslint) task

Running "eslint:tests" (eslint) task

Running "jsonlint:app" (jsonlint) task
>> 3 files lint free.

Running "amdcheck:tests" (amdcheck) task

Total unused paths: 0 in 0 files.
Total unused dependencies: 0 in 0 files.
Total processed files: 73

Running "sasslint:sass" (sasslint) task

Done.


Execution Time (2017-02-06 14:45:06 UTC-0)
loading tasks    2.5s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 20%
eslint:app       6.5s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 54%
eslint:grunt    374ms  ▇▇▇ 3%
eslint:server   391ms  ▇▇▇ 3%
eslint:tests     1.3s  ▇▇▇▇▇▇▇▇ 11%
jsonlint:app    335ms  ▇▇ 3%
amdcheck:tests  262ms  ▇▇ 2%
sasslint:sass   441ms  ▇▇▇ 4%
Total 12.1s

➜  fxa-content-server git:(master)                                                  git:(master|)
```

However, with the `no-nested-ternary`, we do not. :(

```
'no-nested-ternary': 'error',
```

```bash
➜  fxa-content-server git:(master) grunt lint                                       git:(master|)
Running "eslint:app" (eslint) task

Running "eslint:grunt" (eslint) task

Running "eslint:scripts" (eslint) task

Running "eslint:server" (eslint) task

Running "eslint:tests" (eslint) task

/Users/shanetomlinson/development/fxa-content-server/tests/functional/fx_fennec_v1_force_auth.js
  30:29  error  Do not nest ternary expressions  no-nested-ternary

/Users/shanetomlinson/development/fxa-content-server/tests/functional/fx_ios_v1_sign_in.js
  46:29  error  Do not nest ternary expressions  no-nested-ternary

/Users/shanetomlinson/development/fxa-content-server/tests/functional/sync_force_auth.js
  33:29  error  Do not nest ternary expressions  no-nested-ternary

/Users/shanetomlinson/development/fxa-content-server/tests/functional/sync_sign_in.js
  42:29  error  Do not nest ternary expressions  no-nested-ternary

/Users/shanetomlinson/development/fxa-content-server/tests/functional/sync_v2_force_auth.js
  41:29  error  Do not nest ternary expressions  no-nested-ternary

/Users/shanetomlinson/development/fxa-content-server/tests/functional/sync_v2_sign_in.js
  37:29  error  Do not nest ternary expressions  no-nested-ternary

/Users/shanetomlinson/development/fxa-content-server/tests/functional/sync_v3_sign_in.js
  41:29  error  Do not nest ternary expressions  no-nested-ternary

✖ 7 problems (7 errors, 0 warnings)

Warning: Task "eslint:tests" failed. Use --force to continue.

Aborted due to warnings.


Execution Time (2017-02-06 14:46:47 UTC-0)
loading tasks   2.4s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 22%
eslint:app      6.4s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 59%
eslint:grunt   376ms  ▇▇▇ 3%
eslint:server  402ms  ▇▇▇ 4%
eslint:tests    1.3s  ▇▇▇▇▇▇▇▇ 12%
Total 10.9s

➜  fxa-content-server git:(master)
```

Comment (divyabiyani) [inline lib/crypto/butil.js]: Ok @vladikoff , @shane-tomlinson I have opened a new PR w.r.t these changes both in eslint-plugin-fxa and fxa-auth-server.

Comment (shane-tomlinson) [inline lib/crypto/butil.js]: @divyabiyani - ISTM there are multiple PRs to be made out of this one. The first PR is to use the `no new buffer` eslint rule and update any place that's required. 

The 2nd is the more contentious issue of no-space in unary operators except `!` and `!!`, which should probably be discussed with the group before merging.

**Pull request closed without merging**